File: bun-mqtt-backend/tests/dedup.test.ts
import { describe, expect, test } from "bun:test";
import {
  MessageDeduplicator,
  computeDedupKey,
} from "../src/dedup";

describe("MessageDeduplicator", () => {
  test("首次出现返回 true（应处理）", () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 60_000 });
    expect(d.checkAndRecord("k1")).toBe(true);
  });

  test("第二次相同 key 返回 false（重复）", () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 60_000 });
    expect(d.checkAndRecord("k1")).toBe(true);
    expect(d.checkAndRecord("k1")).toBe(false);
    expect(d.checkAndRecord("k1")).toBe(false);
  });

  test("不同 key 互不干扰", () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 60_000 });
    expect(d.checkAndRecord("k1")).toBe(true);
    expect(d.checkAndRecord("k2")).toBe(true);
    expect(d.checkAndRecord("k3")).toBe(true);
    expect(d.checkAndRecord("k1")).toBe(false);
    expect(d.checkAndRecord("k2")).toBe(false);
  });

  test("达到 maxSize 时按 LRU 淘汰最旧条目", () => {
    const d = new MessageDeduplicator({ maxSize: 3, ttlMs: 60_000 });
    d.checkAndRecord("k1");
    d.checkAndRecord("k2");
    d.checkAndRecord("k3");
    // 缓存已满，插入 k4 时应淘汰 k1
    expect(d.checkAndRecord("k4")).toBe(true);
    // k1 已被淘汰，再来一次应视为新消息
    expect(d.checkAndRecord("k1")).toBe(true);
    // k2 应该也被淘汰了（因为 k1 淘汰后 k2 变最旧，插入 k1 又淘汰 k2）
    expect(d.checkAndRecord("k2")).toBe(true);
  });

  test("命中会刷新 LRU 位置（近期使用的不会被淘汰）", () => {
    const d = new MessageDeduplicator({ maxSize: 3, ttlMs: 60_000 });
    d.checkAndRecord("k1");
    d.checkAndRecord("k2");
    d.checkAndRecord("k3");
    // 命中 k1 → k1 变成最近使用
    expect(d.checkAndRecord("k1")).toBe(false);
    // 插入 k4 → 淘汰 k2（现在最旧）
    d.checkAndRecord("k4");
    // k1 应该还在缓存里
    expect(d.checkAndRecord("k1")).toBe(false);
    // k2 应该已被淘汰
    expect(d.checkAndRecord("k2")).toBe(true);
  });

  test("TTL 过期后同 key 视为新消息", async () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 50 });
    expect(d.checkAndRecord("k1")).toBe(true);
    expect(d.checkAndRecord("k1")).toBe(false);
    await new Promise((r) => setTimeout(r, 80));
    expect(d.checkAndRecord("k1")).toBe(true);
  });

  test("disabled 时永远返回 true", () => {
    const d = new MessageDeduplicator({ enabled: false });
    expect(d.checkAndRecord("k1")).toBe(true);
    expect(d.checkAndRecord("k1")).toBe(true);
    expect(d.checkAndRecord("k1")).toBe(true);
  });

  test("sweep() 清理过期条目", async () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 30 });
    d.checkAndRecord("k1");
    d.checkAndRecord("k2");
    expect(d.getStats().size).toBe(2);
    await new Promise((r) => setTimeout(r, 60));
    const removed = d.sweep();
    expect(removed).toBe(2);
    expect(d.getStats().size).toBe(0);
  });

  test("getStats 返回正确的命中率", () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 60_000 });
    d.checkAndRecord("k1");  // miss
    d.checkAndRecord("k1");  // hit
    d.checkAndRecord("k1");  // hit
    d.checkAndRecord("k2");  // miss
    const s = d.getStats();
    expect(s.hits).toBe(2);
    expect(s.misses).toBe(2);
    expect(s.hitRate).toBeCloseTo(0.5);
  });

  test("clear() 重置所有状态", () => {
    const d = new MessageDeduplicator({ maxSize: 100, ttlMs: 60_000 });
    d.checkAndRecord("k1");
    d.checkAndRecord("k1");
    d.clear();
    const s = d.getStats();
    expect(s.size).toBe(0);
    expect(s.hits).toBe(0);
    expect(s.misses).toBe(0);
    // clear 之后同 key 又变成新消息
    expect(d.checkAndRecord("k1")).toBe(true);
  });
});

describe("computeDedupKey", () => {
  test("优先使用 messageId 字段", () => {
    const r = computeDedupKey("t", "raw", { messageId: "abc123" });
    expect(r.source).toBe("id");
    expect(r.key).toBe("id:messageId:abc123");
  });

  test("按优先级尝试多个 ID 字段", () => {
    expect(computeDedupKey("t", "r", { msgId: "x" }).key).toBe("id:msgId:x");
    expect(computeDedupKey("t", "r", { uuid: "u" }).key).toBe("id:uuid:u");
    expect(computeDedupKey("t", "r", { id: 42 }).key).toBe("id:id:42");
    expect(computeDedupKey("t", "r", { message_id: "m" }).key).toBe("id:message_id:m");
  });

  test("messageId 优先于 msgId", () => {
    const r = computeDedupKey("t", "r", { msgId: "b", messageId: "a" });
    expect(r.key).toBe("id:messageId:a");
  });

  test("无 ID 字段时回退到 hash", () => {
    const r = computeDedupKey("sensors/x", '{"temperature":25}', { temperature: 25 });
    expect(r.source).toBe("hash");
    expect(r.key.startsWith("h:")).toBe(true);
  });

  test("相同 topic+raw 产生相同 hash key（幂等）", () => {
    const k1 = computeDedupKey("t", "same-raw", { foo: 1 }).key;
    const k2 = computeDedupKey("t", "same-raw", { foo: 1 }).key;
    expect(k1).toBe(k2);
  });

  test("不同 topic 或 raw 产生不同 hash key", () => {
    const k1 = computeDedupKey("t1", "raw", {}).key;
    const k2 = computeDedupKey("t2", "raw", {}).key;
    const k3 = computeDedupKey("t1", "raw2", {}).key;
    expect(k1).not.toBe(k2);
    expect(k1).not.toBe(k3);
  });

  test("空对象 / 非对象 payload 走 hash", () => {
    expect(computeDedupKey("t", "r", null).source).toBe("hash");
    expect(computeDedupKey("t", "r", undefined).source).toBe("hash");
    expect(computeDedupKey("t", "r", "string").source).toBe("hash");
    expect(computeDedupKey("t", "r", 42).source).toBe("hash");
  });

  test("空字符串 ID 会被忽略，回退到 hash", () => {
    const r = computeDedupKey("t", "r", { messageId: "" });
    expect(r.source).toBe("hash");
  });
});

File: go-mqtt-backend/internal/dedup/dedup_test.go
package dedup

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

// ---------- MessageDeduplicator ----------

func TestDedup_FirstSeenReturnsTrue(t *testing.T) {
	d := New(true, 100, 60_000)
	defer d.Stop()
	assert.True(t, d.CheckAndRecord("k1"))
}

func TestDedup_SecondSameKeyReturnsFalse(t *testing.T) {
	d := New(true, 100, 60_000)
	defer d.Stop()
	assert.True(t, d.CheckAndRecord("k1"))
	assert.False(t, d.CheckAndRecord("k1"))
	assert.False(t, d.CheckAndRecord("k1"))
}

func TestDedup_DifferentKeysIndependent(t *testing.T) {
	d := New(true, 100, 60_000)
	defer d.Stop()
	assert.True(t, d.CheckAndRecord("k1"))
	assert.True(t, d.CheckAndRecord("k2"))
	assert.True(t, d.CheckAndRecord("k3"))
	assert.False(t, d.CheckAndRecord("k1"))
	assert.False(t, d.CheckAndRecord("k2"))
}

func TestDedup_LRUEviction(t *testing.T) {
	d := New(true, 3, 60_000)
	defer d.Stop()
	d.CheckAndRecord("k1")
	d.CheckAndRecord("k2")
	d.CheckAndRecord("k3")
	// 缓存已满，插入 k4 淘汰 k1
	assert.True(t, d.CheckAndRecord("k4"))
	// k1 已被淘汰 → 新消息
	assert.True(t, d.CheckAndRecord("k1"))
	// k2 也被淘汰
	assert.True(t, d.CheckAndRecord("k2"))
}

func TestDedup_HitRefreshesRecency(t *testing.T) {
	d := New(true, 3, 60_000)
	defer d.Stop()
	d.CheckAndRecord("k1")
	d.CheckAndRecord("k2")
	d.CheckAndRecord("k3")
	// 命中 k1 → k1 变最近使用
	assert.False(t, d.CheckAndRecord("k1"))
	// 插入 k4 → 淘汰 k2（现在最旧）
	d.CheckAndRecord("k4")
	// k1 仍在缓存
	assert.False(t, d.CheckAndRecord("k1"))
	// k2 已被淘汰
	assert.True(t, d.CheckAndRecord("k2"))
}

func TestDedup_TTLExpiry(t *testing.T) {
	d := New(true, 100, 50)
	defer d.Stop()
	assert.True(t, d.CheckAndRecord("k1"))
	assert.False(t, d.CheckAndRecord("k1"))
	time.Sleep(80 * time.Millisecond)
	assert.True(t, d.CheckAndRecord("k1"))
}

func TestDedup_DisabledAlwaysTrue(t *testing.T) {
	d := New(false, 100, 60_000)
	defer d.Stop()
	assert.True(t, d.CheckAndRecord("k1"))
	assert.True(t, d.CheckAndRecord("k1"))
	assert.True(t, d.CheckAndRecord("k1"))
}

func TestDedup_SweepRemovesExpired(t *testing.T) {
	d := New(true, 100, 30)
	defer d.Stop()
	d.CheckAndRecord("k1")
	d.CheckAndRecord("k2")
	assert.Equal(t, 2, d.GetStats().Size)
	time.Sleep(60 * time.Millisecond)
	removed := d.sweep()
	assert.Equal(t, 2, removed)
	assert.Equal(t, 0, d.GetStats().Size)
}

func TestDedup_StatsHitRate(t *testing.T) {
	d := New(true, 100, 60_000)
	defer d.Stop()
	d.CheckAndRecord("k1") // miss
	d.CheckAndRecord("k1") // hit
	d.CheckAndRecord("k1") // hit
	d.CheckAndRecord("k2") // miss
	s := d.GetStats()
	assert.Equal(t, int64(2), s.Hits)
	assert.Equal(t, int64(2), s.Misses)
	assert.InDelta(t, 0.5, s.HitRate, 1e-9)
}

func TestDedup_Clear(t *testing.T) {
	d := New(true, 100, 60_000)
	defer d.Stop()
	d.CheckAndRecord("k1")
	d.CheckAndRecord("k1")
	d.Clear()
	s := d.GetStats()
	assert.Equal(t, 0, s.Size)
	assert.Equal(t, int64(0), s.Hits)
	assert.Equal(t, int64(0), s.Misses)
	assert.True(t, d.CheckAndRecord("k1"))
}

// ---------- ComputeDedupKey ----------

func TestComputeDedupKey_PreferMessageID(t *testing.T) {
	r := ComputeDedupKey("t", []byte("raw"), map[string]any{"messageId": "abc123"})
	assert.Equal(t, "id", r.Source)
	assert.Equal(t, "id:messageId:abc123", r.Key)
}

func TestComputeDedupKey_MultipleIDFields(t *testing.T) {
	assert.Equal(t, "id:msgId:x", ComputeDedupKey("t", []byte("r"), map[string]any{"msgId": "x"}).Key)
	assert.Equal(t, "id:uuid:u", ComputeDedupKey("t", []byte("r"), map[string]any{"uuid": "u"}).Key)
	assert.Equal(t, "id:id:42", ComputeDedupKey("t", []byte("r"), map[string]any{"id": float64(42)}).Key)
	assert.Equal(t, "id:message_id:m", ComputeDedupKey("t", []byte("r"), map[string]any{"message_id": "m"}).Key)
}

func TestComputeDedupKey_MessageIDPriorityOverMsgID(t *testing.T) {
	r := ComputeDedupKey("t", []byte("r"), map[string]any{"msgId": "b", "messageId": "a"})
	assert.Equal(t, "id:messageId:a", r.Key)
}

func TestComputeDedupKey_FallbackToHash(t *testing.T) {
	r := ComputeDedupKey("sensors/x", []byte(`{"temperature":25}`), map[string]any{"temperature": float64(25)})
	assert.Equal(t, "hash", r.Source)
	assert.True(t, len(r.Key) > 2 && r.Key[:2] == "h:")
}

func TestComputeDedupKey_HashIdempotent(t *testing.T) {
	k1 := ComputeDedupKey("t", []byte("same-raw"), map[string]any{"foo": float64(1)}).Key
	k2 := ComputeDedupKey("t", []byte("same-raw"), map[string]any{"foo": float64(1)}).Key
	assert.Equal(t, k1, k2)
}

func TestComputeDedupKey_DifferentTopicOrRaw(t *testing.T) {
	k1 := ComputeDedupKey("t1", []byte("raw"), map[string]any{}).Key
	k2 := ComputeDedupKey("t2", []byte("raw"), map[string]any{}).Key
	k3 := ComputeDedupKey("t1", []byte("raw2"), map[string]any{}).Key
	assert.NotEqual(t, k1, k2)
	assert.NotEqual(t, k1, k3)
}

func TestComputeDedupKey_NilOrEmptyPayloadUsesHash(t *testing.T) {
	assert.Equal(t, "hash", ComputeDedupKey("t", []byte("r"), nil).Source)
	assert.Equal(t, "hash", ComputeDedupKey("t", []byte("r"), map[string]any{}).Source)
}

func TestComputeDedupKey_EmptyStringIDIgnored(t *testing.T) {
	r := ComputeDedupKey("t", []byte("r"), map[string]any{"messageId": ""})
	assert.Equal(t, "hash", r.Source)
}
